File: frontend/scripts/validation.js
export function validateUsername(username) {
    if (!username) {
        return "Username is required";
    }
    if (username.length < 3 || username.length > 20) {
        return "Username must be 3–20 characters";
    }
    if (!/(?=.*[a-zA-Z])/.test(username)) {
        return "Username must contain at least one letter";
    }
    if (!/^[a-zA-Z0-9_]+$/.test(username)) {
        return "Only letters, numbers, and underscores allowed";
    }
    if (/^_|_$/.test(username)) {
        return "Username cannot start or end with underscore";
    }

    return null;
}

export function validateEmail(email) {
    if (!email) return "Email is required";

    if (email.includes(" "))
        return "Email cannot contain spaces";

    const emailRegex =
        /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;

    if (!emailRegex.test(email))
        return "Invalid email format";

    return null;
}

File: backend/utils/validation.go
/*
package utils

import "regexp"

	func ValidateUsername(username string) string {
		if username == "" {
			return "Username is required"
		}

		if len(username) < 3 || len(username) > 20 {
			return "Username must be 3–20 characters"
		}

		hasLetter := regexp.MustCompile(`[a-zA-Z]`)
		if !hasLetter.MatchString(username) {
			return "Username must contain at least one letter"
		}

		validCharacters := regexp.MustCompile(`^[a-zA-Z0-9_]+$`)
		if !validCharacters.MatchString(username) {
			return "Only letters, numbers, and underscores allowed"
		}

		if regexp.MustCompile(`^_|_$`).MatchString(username) {
			return "Username cannot start or end with underscore"
		}

		return ""
	}

	func ValidateEmail(email string) string {
		if email == "" {
			return "Email is required"
		}

		if regexp.MustCompile(`\s`).MatchString(email) {
			return "Email cannot contain spaces"
		}

		emailRegex := regexp.MustCompile(
			`^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$`,
		)

		if !emailRegex.MatchString(email) {
			return "Invalid email format"
		}

		return ""
	}
*/
package utils

import (
	"regexp"

	"github.com/go-playground/validator/v10"
)

var Validate *validator.Validate

func InitValidator() {
	Validate = validator.New()

	Validate.RegisterValidation("hasletter", func(fl validator.FieldLevel) bool {
		value := fl.Field().String()
		hasLetter := regexp.MustCompile(`[a-zA-Z]`)
		return hasLetter.MatchString(value)
	})
}
